engine/common/synchronization/finalized_snapshot.go

```diff
@@ -0,0 +1,125 @@
+package synchronization
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/rs/zerolog"
+
+	"github.com/onflow/flow-go/engine"
+	"github.com/onflow/flow-go/model/flow"
+	"github.com/onflow/flow-go/module/lifecycle"
+	"github.com/onflow/flow-go/state/protocol"
+)
+
+// finalizedSnapshot is a helper structure which contains latest finalized header and participants list
+type finalizedSnapshot struct {
+	head         *flow.Header
+	participants flow.IdentityList
+}
+
+// FinalizedSnapshotCache represents a cached snapshot of the latest finalized header and participants list.
+// It is used in Engine to access latest valid data.
+type FinalizedSnapshotCache struct {
+	mu sync.RWMutex
+
+	log                       zerolog.Logger
+	state                     protocol.State
+	identityFilter            flow.IdentityFilter
+	lastFinalizedSnapshot     *finalizedSnapshot
+	finalizationEventNotifier engine.Notifier // notifier for finalization events
+
+	lm *lifecycle.LifecycleManager
+}
+
+// NewFinalizedSnapshotCache creates a new finalized snapshot cache.
+func NewFinalizedSnapshotCache(log zerolog.Logger, state protocol.State, participantsFilter flow.IdentityFilter) (*FinalizedSnapshotCache, error) {
+
+	cache := &FinalizedSnapshotCache{
+		state:          state,
+		identityFilter: participantsFilter,
+		lm:             lifecycle.NewLifecycleManager(),
+		log:            log.With().Str("component", "finalized_snapshot_cache").Logger(),
+	}
+
+	err := cache.updateSnapshot()
+	if err != nil {
+		return nil, fmt.Errorf("could not apply last finalized state")
+	}
+
+	return cache, nil
+}
+
+// get returns last locally stored snapshot which contains final header
+// and list of filtered identities
+func (f *FinalizedSnapshotCache) get() *finalizedSnapshot {
```

Comment:
I considered this at first, but I decided against it because the `get` method is only used with the `synchronization` package, and nowhere else. Therefore, there is no need to expose it as a public method. Even if we did expose it as a public method, it would be useless to callers outside the package because the returned struct is a private struct which only has private fields.